torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,289 @@
+      else:
+        logging.info("XLA Backend: Skipping tensor CC op wait.")
+    else:
+      if distutils.util.strtobool(
+          os.environ.get('XLA_BACKEND_BLOCKING_BARRIER', 'False')):
+        logging.info("XLA Backend: Non-tensor wait...")
+        torch_xla._XLAC._xla_step_marker(
+            torch_xla._XLAC._xla_get_default_device(), devices=[], wait=True)
+      else:
+        logging.info("XLA Backend: Skipping non-tensor wait.")
+
+
+# -------------------------------------
+# Override torch.distributed.new_group.
+# -------------------------------------
+_orig_new_group_fn = dist.new_group
+
+
+def infer_mesh(slice_ranks, world_size):
+  '''Infers a rectangular mesh topology from a slice in the mesh.
+
+    Example, given world size 12 and a slice like the following:
+        [1, 5, 9]
+    this function infers the following mesh:
+        [[0, 4, 8],
+         [1, 5, 9],
+         [2, 6, 10],
+         [3, 7, 11]]
+
+    We only support ractangular meshes.
+    '''
+  slice_len = len(slice_ranks)
+  if world_size % slice_len != 0:
+    raise ValueError(
+        'Given slice length doesn\'t equal to a side of a rectangle. '
+        f'World size: {world_size}, slice length: {slice_len}')
+
+  # ensure the ranks are a correct range.
+  start = slice_ranks[0]
+  step = slice_ranks[1] - slice_ranks[0]
+  stop = start + step * slice_len
+  expected_ranks = list(range(start, stop, step))
+  if slice_ranks != expected_ranks:
+    raise ValueError(
+        f'Given slice isn\'t a range with evenly distributed steps: {slice_ranks}'
+    )
+
+  num_slices = int(world_size / slice_len)
+  world = list(range(world_size))
+  mesh = []
+  if step == 1:
+    # Horizontal case.
+    if start % slice_len != 0 or start >= world_size:
+      raise ValueError('Given horizontal slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i * slice_len
+      slice_stop = (i + 1) * slice_len
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  else:
+    # Vertical case.
+    if start >= num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct start: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    if step != num_slices:
+      raise ValueError('Given vertical slice doesn\'t have a correct step: '
+                       f'World size: {world_size}, slice length: {slice_len}, '
+                       f'slice: {slice_ranks}')
+    for i in range(num_slices):
+      slice_start = i
+      slice_stop = i + slice_len * step
+      slice_idx = slice(slice_start, slice_stop, step)
+      mesh.append(world[slice_idx])
+  assert slice_ranks in mesh
+
+  return mesh
+
+
+def new_xla_process_group(ranks=None,
```

Comment:
This seems to be a new user-facing API. Do we need API docs explain the expectations on arguments (especially the differences with c10d `new_group`)?

Reply:
I think there is no new expectation on arguments? Did you see any?

Reply:
the original `new_group` expects that `ranks` is either `None` or a `list` of `int`. Looks like it now can also accept `List[List[Integer]]`?

Reply:
I don't think `ranks` can be a list of list. Can you point me to the line of code that expect that?

The **output** `_mesh` will be a list of list (a rectangular 2-d array).

Reply:
My bad, it's not 2d array.